### PR TITLE
Transfer leadership when establishLeadership fails

### PR DESCRIPTION
**Pull request by i0rek**
Fixes https://github.com/hashicorp/consul/issues/5047. I am throwing this out here to have a place to discuss the approaches. 

__TLDR;__ transition leadership is now implemented in raft which is being used in this PR. 

Background:

Leadership is established in raft and there is basically no way to influence that from consul. And there has been no need to. But when we setup leadership in consul in `establishLeadership` things can go wrong and in the past we retried to fix it. This is problematic since even though raft considers this server the leader, it is not fully prepared! Consistent reads for example are not possible.

Abandoned Ideas:

* calling `revokeLeadership`: that tears down some consul things, but has no influence on raft leadership
* quitting `leaderLoop`: that resets the consul side of things, but doesn't change the raft leadership

Solutions:

* this PR: calling `DemoteVoter` as the leader, will make this server step down by marking the current leader as nonvoter. The same server is promoted to a voter afterwards again by autopilot from the new leader: https://github.com/hashicorp/consul/blob/884b2e00af8bdde14ca5d0f79cde77929e35664b/agent/consul/autopilot/autopilot.go#L304-L329 The idea to that approach came from https://github.com/hashicorp/raft/issues/218.
* alternatively we could implement leadership transfer in raft as specified in https://ramcloud.stanford.edu/~ongaro/thesis.pdf chapter 3.10.

Concerns with demotevoter:

* in an environment where not all servers are running autopilot, it is possible that the leader doesn't run it and doesn't promote the demoted server to be a voter again. Lets assume we would be left with 2 voters and 1 nonvoter. That would make electing a leader harder, but raft would figure it out eventually. During an upgrade, the same could happen one last time: 1 voter and 2 nonvoters left which is the worst case. But when this happens, electing a leader is easy again. As soon as the leader runs autopilot, all nonvoters will be promoted to a voter again. __`establishLeadership` would have to fail often during that process to make this scenario possible__. Without this fix though, we would be stuck with a leader that is not performing all leader tasks, without a chance of recovery.
* raft state is modified for this and we rely on consul for promoting the previous leader to be a voter again - feels fishy.

Concerns with leadership transfer: 

none

This PR uses the leadership transfer feature. It requires a revendoring though and fails because of it.

### CODE REVIEW

File: agent/consul/server.go [hunk 1]
```diff
@@ -540,7 +541,12 @@ func (s *Server) setupRaft() error {
 
 	// Make sure we set the LogOutput.
 	s.config.RaftConfig.LogOutput = s.config.LogOutput
-	s.config.RaftConfig.Logger = s.logger
+	raftLogger := hclog.New(&hclog.LoggerOptions{
+		Name:   "raft",
+		Level:  hclog.LevelFromString(s.config.LogLevel),
+		Output: s.config.LogOutput,
+	})
+	s.config.RaftConfig.Logger = raftLogger
```
Comment (i0rek) [inline agent/consul/server.go]: Raft logging changed and we need to provide the hclogger now.

File: agent/consul/leader.go [hunk 2]
```diff
@@ -178,7 +195,7 @@ RECONCILE:
 			if err := s.revokeLeadership(); err != nil {
 				s.logger.Printf("[ERR] consul: failed to revoke leadership: %v", err)
 			}
-			goto WAIT
+			return err
```
Comment (i0rek) [inline agent/consul/leader.go]: This usecase #1: we couldn't establish leadership and instead of retrying it after `interval` in `WAIT`, an error is return which leads to a raft leadership transfer.

File: agent/consul/leader.go [hunk 3]
```diff
@@ -213,17 +230,21 @@ WAIT:
...
 		case member := <-reconcileCh:
 			s.reconcileMember(member)
 		case index := <-s.tombstoneGC.ExpireCh():
 			go s.reapTombstones(index)
 		case errCh := <-s.reassertLeaderCh:
-			errCh <- reassert()
+			err := reassert()
+			errCh <- err
+			if err != nil {
+				return err
```
Comment (i0rek) [inline agent/consul/leader.go]: This is the second usecase: we couldn't reassert and instead of retrying it after `interval`, an error is returned which leads to a raft leadership transfer.

File: agent/consul/leader.go [hunk 4]
```diff
@@ -126,9 +129,23 @@ func (s *Server) monitorLeadership() {
 	}
 }
 
+func (s *Server) leadershipTransfer() {
+	retryCount := 3
+	for i := 0; i < retryCount; i++ {
```
Comment (freddygv) [inline agent/consul/leader.go]: Should this function take some action if the leadershipTransfer is unsuccessful after the 3 tries?

File: agent/consul/leader.go [hunk 5]
```diff
@@ -213,17 +230,21 @@ WAIT:
...
 		case <-interval:
 			goto RECONCILE
 		case member := <-reconcileCh:
 			s.reconcileMember(member)
 		case index := <-s.tombstoneGC.ExpireCh():
 			go s.reapTombstones(index)
 		case errCh := <-s.reassertLeaderCh:
-			errCh <- reassert()
+			err := reassert()
+			errCh <- err
```
Comment (freddygv) [inline agent/consul/leader.go]: How come we're sending an err into the errCh here? Seems like we should just return if `err != nil` if we don't want to retry the reassert. (since this is the only place where we read from that channel)

File: agent/consul/leader.go (hunk 4 above)
Comment (freddygv) [reply agent/consul/leader.go]: Maybe the `weAreLeaderCh` should be closed and set to nil, since if we then receive from `aclUpgradeCh` and that channel is still open, we initialize ACLs when we probably shouldn't.

File: agent/consul/leader.go (hunk 3 above)
Comment (banks) [reply agent/consul/leader.go]: Hmmm, when does this happen? I don't recall the specifics and wonder if by failing and stepping down we might end up causing new leadership stability issues?

File: agent/consul/leader.go [hunk 6]
```diff
@@ -77,7 +77,10 @@ func (s *Server) monitorLeadership() {
 				leaderLoop.Add(1)
 				go func(ch chan struct{}) {
 					defer leaderLoop.Done()
-					s.leaderLoop(ch)
+					err := s.leaderLoop(ch)
+					if err != nil {
+						s.leadershipTransfer()
```
Comment (banks) [inline agent/consul/leader.go]: What if leadership transfer fails? We exit the leader loop anyway and no longer act as the leader, but does raft still think we are the leader? That seems like a bad case to be in - roughly the same as the bug this is meant to be fixing although we at least made an attempt at telling raft we didn't want to be leader any more.

I can think of a couple of possible options here:
 - Make a method in Raft lib that allows the leader to "StepDown" - it will stop running heartbeats etc. but not stop being a voter - basically, force a leader into follower state.
 - Keep trying the leadership Transfer indefinitely because the cluster is broken until it works anyway.
 - Keep retrying for a limited length of time and then crash the whole process to force a step down.

**Comment by banks** [comment]
@i0rek could you also double check how the `hclog` integration pans out. It came up today that Vault and Nomad have different date formats in logs which tripped someone up. Can you make sure we don't make it even worse by having raft log lines log with a different date format to the rest of Consul's logs please? I'm not sure if that is the case or not as I didn't look hard at where the plumbing/config sits for that but would just like to sanity check it.

File: agent/consul/leader.go (hunk 6 above)
Comment (i0rek) [reply agent/consul/leader.go]: I made a couple of changes to my PR, so that `leaderLoop` is only left if `leadershipTransfer` was successful. I think that mitigates the issues you are talking about.

What happens now is that in case `leadershipTransfer` fails, consul stays in `leaderLoop` and waits until `ReconcileInterval` to retry to `establishLeadership` (and to `transferLeadership` in case it fails).

File: agent/consul/leader.go (hunk 4 above)
Comment (i0rek) [reply agent/consul/leader.go]: I changed my PR a bit, `leaderLoop` is only left if `transferLeadership` was successful. I think that answers your question, because `leaderLoop` will retry `transferLeadership` if it failed before.

File: agent/consul/leader.go (hunk 5 above)
Comment (i0rek) [reply agent/consul/leader.go]: This channel is also read by the snapshot endpoint. It waits for reassertion to be successful before reporting that back to the caller. This is why we need to send our errors (or `nil`) there.

File: agent/consul/leader.go [hunk 7]
```diff
@@ -223,7 +229,39 @@ WAIT:
...
+			err := s.establishLeadership()
+			errCh <- err
+
+			// in case reassert failed, which means that
+			// establishLeadership failed, we will try to transfer
+			// leadership. At this time raft thinks we are the
+			// leader, but consul disagrees.
+			if err != nil {
+				if err := s.leadershipTransfer(); err != nil {
+					goto WAIT
```
Comment (banks) [inline agent/consul/leader.go]: So now we attempt to transfer 3 times but if it fails we still hang out in non-leader limbo land for a bit before retrying?

I guess this is what i mentioned as "retry indefinitiely" and it should really work immediately if rest of the cluster is in an OK state so I think this is good.

File: agent/consul/leader.go (hunk 6 above)
Comment (banks) [reply agent/consul/leader.go]: I think this is OK although `ReconcileInterval` is 60 seconds which seems a lot to wait around for when we know the cluster is down. 🤔

Should we instead consider looping indefinitely retrying every 5 seconds or something?

I think this is way better than before and eventually it should recover so it's good just wondering if it's easy to make that better quicker?

File: agent/consul/leader.go (hunk 7 above)
Comment (i0rek) [reply agent/consul/leader.go]: My thinking was that since leadershipTransfer failed, we try to establishLeadership again. I think in general establishLeadership is more likely to succeed than transferLeadership. I think if I make the interval smaller - like 5 seconds - before it retries it is the better solution than trying to transfer leadership indefinitely.

File: agent/consul/leader.go (hunk 3 above)
Comment (i0rek) [reply agent/consul/leader.go]: https://github.com/hashicorp/consul/blob/09a1a32cfd57cfb83c08f0d7d89e40a92a06a138/agent/consul/server.go#L245-L246

This happens after snapshot restore. Since the agent revokes leadership and immediately tries to establish it again, there is the possibility that it fails. When it does we are in the same situation as above - raft thinks this agent is a leader, but consul disagrees.

File: agent/consul/leader.go (hunk 6 above)
Comment (i0rek) [reply agent/consul/leader.go]: I reset the `interval` to 5 seconds when transfer leadership fails so that it retries establishLeadership again faster than before.

**Comment by i0rek** [comment]
@banks I checked the logs and fixed the timestamps, but sometimes there are two whitespaces before raft:

master:
```
2019/06/13 22:11:45 [INFO] agent: Started DNS server 127.0.0.1:8600 (tcp)
2019/06/13 22:11:45 [INFO] agent: Started DNS server 127.0.0.1:8600 (udp)
2019/06/13 22:11:45 [INFO] agent: Started HTTP server on 127.0.0.1:8500 (tcp)
2019/06/13 22:11:45 [INFO] agent: started state syncer
2019/06/13 22:11:45 [INFO] agent: Started gRPC server on 127.0.0.1:8502 (tcp)
2019/06/13 22:11:45 [WARN] raft: Heartbeat timeout from "" reached, starting election
2019/06/13 22:11:45 [INFO] raft: Node at 127.0.0.1:8300 [Candidate] entering Candidate state in term 2
2019/06/13 22:11:45 [DEBUG] raft: Votes needed: 1
2019/06/13 22:11:45 [DEBUG] raft: Vote granted from 2ef30f59-536b-3efa-80d8-5bd3afb67585 in term 2. Tally: 1
2019/06/13 22:11:45 [INFO] raft: Election won. Tally: 1
```

pr:
```
2019/06/13 22:10:58 [INFO] agent: Started DNS server 127.0.0.1:8600 (tcp)
2019/06/13 22:10:58 [INFO] agent: Started DNS server 127.0.0.1:8600 (udp)
2019/06/13 22:10:58 [INFO] agent: Started HTTP server on 127.0.0.1:8500 (tcp)
2019/06/13 22:10:58 [INFO] agent: started state syncer
2019/06/13 22:10:58 [INFO] agent: Started gRPC server on 127.0.0.1:8502 (tcp)
2019/06/13 22:10:58 [WARN]  raft: Heartbeat timeout from "" reached, starting election
2019/06/13 22:10:58 [INFO]  raft: Node at 127.0.0.1:8300 [Candidate] entering Candidate state in term 2
2019/06/13 22:10:58 [DEBUG] raft: Votes needed: 1
2019/06/13 22:10:58 [DEBUG] raft: Vote granted from bd52484a-2349-6478-2e1f-4eb4c64c924c in term 2. Tally: 1
2019/06/13 22:10:58 [INFO]  raft: Election won. Tally: 1
```

File: agent/consul/leader.go [hunk 8]
```diff
@@ -223,7 +230,47 @@ WAIT:
...
+				if err := s.leadershipTransfer(); err != nil {
+					// establishedLeader was true before,
+					// but it no longer is since it revoked
+					// leadership and Leadership transfer
+					// also failed. Which is why it stays
+					// in the leaderLoop, but now
+					// establishedLeader needs to be set to
+					// false.
+					establishedLeader = false
+					interval = time.After(5 * time.Second)
```
Comment (banks) [inline agent/consul/leader.go]: I think this is OK because we'll wait for 5 seconds in next wait loop and then timeout and `goto RECONCILE`. Right after that label we will hit `interval := time.After(s.config.ReconcileInterval)` again which _presumably_ will set the interval back again.

I must admit I'm a little confused about how `:=` works after a GOTO - it's reassigning the same variable name but in the same scope on subsequent jumps? I wonder if this is some strange variant of variable shadowing even though they are in the same scope? Maybe Go just has a special case to allow this when using GOTO but not in serial code? If it works I guess it's fine 😄

Comment (i0rek) [reply agent/consul/leader.go]: I am not sure what you mean. `WAIT` is well after the `RECONCILE` and the `interval` variable declaration and the code should just be using the same variable. 

I reproed your question in a playground: https://play.golang.org/p/6AqssHXg3Wt. If you jump before a variable declaration golang will create a new variable. Or did you mean something else?

Comment (banks) [reply agent/consul/leader.go]: Yeah I did a similar repro and it's fine, just was a strange one.

The path here is:

 - we set interval to be a timer chan that will go off in 5 seconds
 - we `goto WAIT` which enters a select on a few things including that chan
 - when the chan fires, that select branch does `goto RECONCILE` which immeidately re-assigns a timer chan for the original ReconcileInterval (using `:=`).

My concern was that we might end up regaining leadership and then doing reconcile every 5 seconds after that but it's not the case due to the path mentioned above.

It also occurs to me that we always have had a re-assignment after `goto RECONCILE` so it's not really any different than before, it's just that was the only assignment before and I wondered if some strange form of shadowing might cause issues. That appears not to be the case so I think this is fine!

Comment (banks) [reply agent/consul/leader.go]: For fun - Go does create a new variable with the same name.

You can see that here: https://play.golang.org/p/FU0ZxictDXE capturing the variable in a lambda and then looping with GOTO leaves the lamda holding the original value not the redefined one after the GOTO jump.

It's just weird to me because it's in the same scope - shadowing across scopes seems fine but this seems to be a special case you can't normally do outside of GOTO.

Comment (i0rek) [reply agent/consul/leader.go]: wow, now I see what you mean.

**Comment by silenceper** [comment]
This pr add vendoring of github.com/hashicorp/consul
conflict with #5943

**Comment by mkeeler** [comment]
@slienceper Looks like you are correct? One of the other engineers has been working on a circle ci check to ensure we aren't accidentally doing this (which appears to be a common mistake for us).
